openpype/hosts/max/plugins/publish/validate_attributes.py

```diff
@@ -0,0 +1,95 @@
+# -*- coding: utf-8 -*-
+"""Validator for Attributes."""
+from pyblish.api import ContextPlugin, ValidatorOrder
+from pymxs import runtime as rt
+
+from openpype.pipeline.publish import (
+    OptionalPyblishPluginMixin,
+    PublishValidationError,
+    RepairContextAction
+)
+
+
+class ValidateAttributes(OptionalPyblishPluginMixin,
+                         ContextPlugin):
+    """Validates attributes are consistent in 3ds max."""
+
+    order = ValidatorOrder
+    hosts = ["max"]
+    label = "Attributes"
+    actions = [RepairContextAction]
+    optional = True
+
+    @classmethod
+    def get_invalid(cls, context):
+        attributes = (
+            context.data["project_settings"]["max"]["publish"]
+                        ["ValidateAttributes"]["attributes"]
+        )
+        if not attributes:
+            return
+
+        for wrap_object, property_name in attributes.items():
```

Comment:
Why is this called a **wrap** object? What does that refer to?
And why is this an extra for loop that we 'return' directly in the iteration as soon an error occurs. Shouldn't we be collecting any errors for all wrap objects?

Isn't it just:
```python


def has_property(object_name, property_name):
    return rt.Execute(f'isProperty {object_name} "{property_name}"')]


def is_matching_value(object_name, property_name, value):
    property_value = rt.Execute(f"{object_name}.{property_name}")
    
    # Wrap property value if value is a string valued attributes
    # starting with a `#`
    if (
        isinstance(value, str) and 
        value.startswith("#") and 
        not value.endswith(")")
    ):
        # prefix value with `#`
        # not applicable for #() array value type
        # and only applicable for enum i.e. #bob, #sally
        property_value = f"#{property_value}"
        
    return property_value == value

invalid = []
for object_name, required_properties in attributes.items();
    node = rt.GetNodeByName(object_name)
    if not node:
        # do nothing if node does not exist?
        continue

    for property_name, value in required_properties.items():
        if not has_property(object_name, property_name):
            cls.log.error(f"Non-existing property: {object_name}.{property_name}")
            invalid.append((object_name, property_name))

        if not is_matching_value(object_name, property_name, value):
            cls.log.error(f"Invalid value for: {object_name}.{property_name}. Should be: {value}")
            invalid.append((object_name, property_name))

return invalid

Also, shouldn't we actually only run it if the object name itself exists?

Reply:
> Why is this called a **wrap** object? What does that refer to? And why is this an extra for loop that we 'return' directly in the iteration as soon an error occurs. Shouldn't we be collecting any errors for all wrap objects?
> 
> Isn't it just:
> 
> ```python
> def has_property(object_name, property_name):
>     """Return whether an object has a property with given name"""
>     return rt.Execute(f'isProperty {object_name} "{property_name}"')]
> 
> 
> def is_matching_value(object_name, property_name, value):
>     """Return whether an existing property matches value `value"""
>     property_value = rt.Execute(f"{object_name}.{property_name}")
>     
>     # Wrap property value if value is a string valued attributes
>     # starting with a `#`
>     if (
>         isinstance(value, str) and 
>         value.startswith("#") and 
>         not value.endswith(")")
>     ):
>         # prefix value with `#`
>         # not applicable for #() array value type
>         # and only applicable for enum i.e. #bob, #sally
>         property_value = f"#{property_value}"
>         
>     return property_value == value
> 
> invalid = []
> for object_name, required_properties in attributes.items();
>     node = rt.GetNodeByName(object_name)
>     if not node:
>         # do nothing if node does not exist?
>         continue
> 
>     for property_name, value in required_properties.items():
>         if not has_property(object_name, property_name):
>             cls.log.error(f"Non-existing property: {object_name}.{property_name}")
>             invalid.append((object_name, property_name))
> 
>         if not is_matching_value(object_name, property_name, value):
>             cls.log.error(f"Invalid value for: {object_name}.{property_name}. Should be: {value}")
>             invalid.append((object_name, property_name))
> 
> return invalid
> 
> Also, shouldn't we actually only run it if the object name itself exists?
> ```

maybe we should call class object instead of wrap object.

Reply:
How about node type, object type or class type if that's what you mean?

Reply:
it is more likely property type, basically attributes like e.g. `separateAovFIles` is the property of `renderers.current`, which is a object from Max Wrapper.

Reply:
So these are not at all related to any node/object inside the workfile, but are global 3dsmax properties in the file? Or what 3DsMax docs seem to refer to as `SystemGlobal` properties?

I see - yeah, not sure what those would be best called.

I think the Validator can definitely have a bit more descriptive docstring as to what this is about. Maybe also showing an example of the property like `renderers.current` or `renderers.production`?

Reply:
> So these are not at all related to any node/object inside the workfile, but are global 3dsmax properties in the file? Or what 3DsMax docs seem to refer to as `SystemGlobal` properties?
> 
> I see - yeah, not sure what those would be best called.
> 
> I think the Validator can definitely have a bit more descriptive docstring as to what this is about. Maybe also showing an example of the property like `renderers.current` or `renderers.production`?

It is not really systemglobal properties at all, these properties' values can be obtained from the external plugins(like tyFlow, redshift etc.)

I agree that we should have the docstring to tell what it is. Maybe we can even write documentation to mark down the usage since the type of properties are varied and users are probably confused about the usage of json in the setting.